### PR TITLE
feat: Simulator API

### PR DESCRIPTION
**Pull request by vifer**
### Added

- Added simulations api [related changelog](https://developer.paddle.com/changelog/2024/webhook-simulator?utm_source=dx&utm_medium=paddle-php-sdk)

### CODE REVIEW

File: src/Entities/SimulationRunEvent.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+use Paddle\SDK\Entities\SimulationRunEvent\SimulationRunEventResponse;
+use Paddle\SDK\Entities\SimulationRunEvent\SimulationRunEventStatus;
+
+class SimulationRunEvent implements Entity
+{
+    private function __construct(
+        public string $id,
+        public SimulationRunEventStatus $status,
+        public EventTypeName $type,
+        public JSONObject $payload,
```
Comment (vifer) [inline src/Entities/SimulationRunEvent.php:19]: Should this be `NotificationEntity`?

File: src/Entities/Simulation.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+class Simulation implements Entity
+{
+    private function __construct(
+        public string $id,
+        public SimulationStatus $status,
+        public string $notificationSettingId,
+        public string $name,
+        public SimulationSingleEventType|SimulationScenarioType $type,
+        public JSONObject|null $payload,
```
Comment (vifer) [inline src/Entities/Simulation.php:20]: Should this be type `NotificationEntity`?

**Review by davidgrayston-paddle — DISMISSED** [review]

File: src/Entities/SimulationRun.php [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        public \DateTimeInterface $updatedAt,
+    ) {
+    }
+
+    public static function from(array $data): self
+    {
+        return new self(
+            id: $data['id'],
+            status: SimulationRunStatus::from($data['status']),
+            type: SimulationSingleEventType::from($data['type'])->isKnown() ? SimulationSingleEventType::from($data['type']) : SimulationScenarioType::from($data['type']),
```
[the remark this answers is not in the document]
Comment (vifer) [reply src/Entities/SimulationRun.php:27]: yes, that was the idea 😄

File: src/Entities/Simulation.php (hunk 2 above)
Comment (mikeymike) [reply src/Entities/Simulation.php:20]: ^ ignore this, I miss remembered the interface usage for `Notifications\Entity` and actually this I think would be fine to use

File: src/Entities/SimulationType.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public static function from(array $data): self
+    {
+        return new self(
+            name: $data['name'],
+            label: $data['label'],
+            description: $data['description'],
+            group: $data['group'],
+            type: SimulationTypeName::from($data['type']),
```
[the remark this answers is not in the document]
Comment (davidgrayston-paddle) [reply src/Entities/SimulationType.php:36]: Node SDK also has this as `SimulationKind` https://github.com/PaddleHQ/paddle-node-sdk/pull/49/files

Currently it is fine in the PHP SD, as it's using `SimulationSingleEventType|SimulationScenarioType`, but if we ever need to consolidate these into a single `SimulationType`, there might be a naming conflict/crossover.

File: src/Resources/SimulationRuns/SimulationRunsClient.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            $parser->getData(),
+            new Paginator($this->client, $parser->getPagination(), SimulationRunCollection::class),
+        );
+    }
+
+    /**
+     * @throws ApiError          On a generic API error
+     * @throws MalformedResponse If the API response was not parsable
+     */
+    public function get(string $simulationId, string $id): SimulationRun
```
[the remark this answers is not in the document]
Comment (vifer) [reply src/Resources/SimulationRuns/SimulationRunsClient.php]: done ✅

File: src/Entities/SimulationType.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (vifer) [reply src/Entities/SimulationType.php]: class name updated ✅

File: src/Entities/SimulationRun.php [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+class SimulationRun implements Entity
+{
+    private function __construct(
+        public string $id,
+        public SimulationRunStatus $status,
+        public SimulationSingleEventType|SimulationScenarioType $type,
+        public \DateTimeInterface $createdAt,
+        public \DateTimeInterface $updatedAt,
+    ) {
+    }
```
[the remark this answers is not in the document]
Comment (vifer) [reply src/Entities/SimulationRun.php]: done ✅

File: src/Entities/Simulation/SimulationSingleEventType.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+ * @method static SimulationSingleEventType TransactionCanceled()
+ * @method static SimulationSingleEventType TransactionCompleted()
+ * @method static SimulationSingleEventType TransactionCreated()
+ * @method static SimulationSingleEventType TransactionPaid()
+ * @method static SimulationSingleEventType TransactionPastDue()
+ * @method static SimulationSingleEventType TransactionPaymentFailed()
+ * @method static SimulationSingleEventType TransactionReady()
+ * @method static SimulationSingleEventType TransactionUpdated()
+ */
+final class SimulationSingleEventType extends PaddleEnum
```
Comment (mikeymike) [inline src/Entities/Simulation/SimulationSingleEventType.php:60]: Can we check if this is the same as `EventTypeName`

Comment (vifer) [reply src/Entities/Simulation/SimulationSingleEventType.php]: Yes, updated and dropped this class as they should match

File: src/Resources/Simulations/Operations/UpdateSimulation.php [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+class UpdateSimulation implements \JsonSerializable
+{
+    use FiltersUndefined;
+
+    public function __construct(
+        public readonly string|Undefined $notificationSettingId = new Undefined(),
+        public readonly SimulationSingleEventType|SimulationScenarioType|Undefined $type = new Undefined(),
+        public readonly string|Undefined $name = new Undefined(),
+        public readonly SimulationStatus|Undefined $status = new Undefined(),
+        public readonly JSONObject|Undefined $payload = new Undefined(),
```
[the remark this answers is not in the document]
Comment (vifer) [reply src/Resources/Simulations/Operations/UpdateSimulation.php]: done ✅

File: src/Entities/Shared/JSONObject.php [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+/**
+ * |------
+ * | ! Generated code !
+ * | Altering this code will result in changes being overwritten |
+ * |-------------------------------------------------------------|.
+ */
+
+namespace Paddle\SDK\Entities\Shared;
+
+class JSONObject implements \JsonSerializable
```
[the remark this answers is not in the document]
Comment (vifer) [reply src/Entities/Shared/JSONObject.php]: No need, deleted

File: examples/simulations.php
Comment (davidgrayston-paddle) [inline examples/simulations.php:1]: @vifer added simulation example to demonstrate payload creation

**Review by davidgrayston-paddle — DISMISSED** [review]
Looks good 👍

Added example to demonstrate simulation create/update

**Review by davidgrayston-paddle — APPROVED** [review]

**Pull request merged**